src/mito2/src/read/dedup.rs

```diff
@@ -0,0 +1,338 @@
+// Copyright 2023 Greptime Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Utilities to remove duplicate rows from a sorted batch.
+
+use async_trait::async_trait;
+use common_telemetry::debug;
+use common_time::Timestamp;
+
+use crate::error::Result;
+use crate::metrics::MERGE_FILTER_ROWS_TOTAL;
+use crate::read::{Batch, BatchReader};
+
+/// A reader that dedup sorted batches from a source based on the
+/// dedup strategy.
+pub(crate) struct DedupReader<R> {
+    source: R,
+    strategy: Box<dyn DedupStrategy>,
```

Comment:
nit: It seems we can use the statical dispatch?

Reply:
Yes, but I hesitated about it.